### ISSUE TITLE
How to use action cable to broadcast notification JSON instead of HTML

### ISSUE THREAD

**Issue by farooqch11** [simukappu/activity_notification#111]
i am integrating this gem in rails as API and facing alot of over rides! 
need help to over ride the action cable method to broadcast JSON

right now its posting HTML

**Comment by simukappu** [comment]
Hi, thank you for your question and feedback.
The most simple way to customize Action Cable response is customizing notification view.
https://github.com/simukappu/activity_notification#notification-views
As you mentioned, default views are HTML format. You can customize it into JSON format representing your notification model as you like.

If you are expecting `activity_notification`'s Action Cable returns API response as gem's default function, please share your expected format with us. Thanks!

**Comment by farooqch11** [comment]
i created a file under config/initializes/activity_notification_ext.rb

and this code! which works fine for me.

`require "activity_notification/apis/notification_api.rb"

ActivityNotification::NotificationApi.module_eval do
  if Rails::VERSION::MAJOR >= 5
    def broadcast_to_action_cable_channel(params = {})
      if target.notification_action_cable_allowed?(notifiable, key) &&
          notifiable.notification_action_cable_allowed?(target, key)
        target_channel_name = "#{ActivityNotification.config.notification_channel_prefix}_#{target_type}#{ActivityNotification.config.composite_key_delimiter}#{target_id}"
        index_options = params.slice(:filter, :limit, :without_grouping, :with_group_members, :filtered_by_type, :filtered_by_group_type, :filtered_by_group_id, :filtered_by_key)
        ActionCable.server.broadcast(target_channel_name,
          JSON.parse(
            ApplicationController.render(
              template: "activity_notification/notifications/customers/_notification.json.jbuilder",
              locals: {notification: ActivityNotification::Notification.find(id)}
            )
          )    
        )
      end
    end
  end
end`

**Comment by simukappu** [comment]
Thank you for your feedback!
Working in progress to create api mode for activity_notification.

**Comment by irondnb** [comment]
I went easy way and implemented broadcasting as optional target before version 2.0
@simukappu are you going to add some way to turn off/on broadcasting with target subscription?

```
class UserSocket < ActivityNotification::OptionalTarget::Base
  def notify(notification, options = {})
    NotificationBroadcastJob.perform_later(notification)
  end
end

class NotificationBroadcastJob < ApplicationJob
  queue_as :default

  def perform(notification)
    ActionCable.server.broadcast(
      "notifications-#{notification.target.id}",
      notification: JSON.parse(notification_details(notification))
    )
  end

  def notification_details(notification)
    ApplicationController.render(
      partial: 'activity_notification/notifications_with_devise/notification',
      locals: { notification: notification }
    )
  end
end
```

**Comment by simukappu** [comment]
Hi @irondnb,
As you mentioned, I am considering moving this internal implementation for ActionCable broadcasting to optional targets. Then, we can create broadcasting returning JSON response or that integrated with authentication.
With broadcasting using optional targets, we can manage subscription by *Subscription#subscribe_to_optional_target* and so on.

**Comment by simukappu** [comment]
Released as v2.1.0. Thanks!
